libyul/optimiser/UnusedStoreEliminator.cpp

```diff
@@ -0,0 +1,404 @@
+{
+	for (auto& [statement, state]: m_stores[YulString{}])
+		if (state == State::Undecided)
+		{
+			Operation const& storeOperation = m_storeOperations.at(statement);
+			if (_operation.effect == Effect::Read && !knownUnrelated(storeOperation, _operation))
+				state = State::Used;
+			else if (_operation.effect == Effect::Write && knownCovered(storeOperation, _operation))
+				state = State::Unused;
+		}
+}
+
+bool UnusedStoreEliminator::knownUnrelated(
+	UnusedStoreEliminator::Operation const& _op1,
+	UnusedStoreEliminator::Operation const& _op2
+) const
+{
+	KnowledgeBase knowledge(m_dialect, m_ssaValues);
+
+	if (_op1.location != _op2.location)
+		return true;
+	if (_op1.location == Location::Storage)
+	{
+		if (_op1.start && _op2.start)
+			return knowledge.knownToBeDifferent(*_op1.start, *_op2.start);
+	}
+	else
+	{
+		yulAssert(_op1.location == Location::Memory, "");
+		if (
+			(_op1.length && knowledge.knownToBeZero(*_op1.length)) ||
+			(_op2.length && knowledge.knownToBeZero(*_op2.length))
+		)
+			return true;
+
+		if (_op1.start && _op1.length && _op2.start)
+		{
+			optional<u256> length1 = knowledge.valueIfKnownConstant(*_op1.length);
+			optional<u256> start1 = knowledge.valueIfKnownConstant(*_op1.start);
+			optional<u256> start2 = knowledge.valueIfKnownConstant(*_op2.start);
+			if (
+				(length1 && start1 && start2) &&
+				*start1 + *length1 >= *start1 && // no overflow
+				*start1 + *length1 <= *start2
+			)
+				return true;
+		}
+		if (_op2.start && _op2.length && _op1.start)
+		{
+			optional<u256> length2 = knowledge.valueIfKnownConstant(*_op2.length);
+			optional<u256> start2 = knowledge.valueIfKnownConstant(*_op2.start);
+			optional<u256> start1 = knowledge.valueIfKnownConstant(*_op1.start);
+			if (
+				(length2 && start2 && start1) &&
+				*start2 + *length2 >= *start2 && // no overflow
+				*start2 + *length2 <= *start1
+			)
+				return true;
+		}
+
+		if (_op1.start && _op1.length && _op2.start && _op2.length)
+		{
+			optional<u256> length1 = knowledge.valueIfKnownConstant(*_op1.length);
+			optional<u256> length2 = knowledge.valueIfKnownConstant(*_op2.length);
+			if (
+				(length1 && *length1 <= 32) &&
+				(length2 && *length2 <= 32) &&
+				knowledge.knownToBeDifferentByAtLeast32(*_op1.start, *_op2.start)
+			)
+				return true;
+		}
+
+		// TODO for more complicated memory overlap rules, we might need to
+		// impose a non-overflow assumption.
+		// We owuld want to cover things like
+		// op1.start + op1.length <= op2.start
+	}
+
+	return false;
+}
+
+bool UnusedStoreEliminator::knownCovered(
+	UnusedStoreEliminator::Operation const& _covered,
+	UnusedStoreEliminator::Operation const& _covering
+) const
+{
+	if (_covered.location != _covering.location)
+		return false;
+	if (
+		(_covered.start && _covered.start == _covering.start) &&
+		// length is unused for storage
+		(_covered.location == Location::Storage || (_covered.length && _covered.length == _covering.length))
```

Comment:
```suggestion
		// length is unused for storage
		(_covered.length && _covered.length == _covering.length)
```
Since they're set to 1 in the other PR now for storage, this can be simplified, can't it?